### ISSUE TITLE
Faster implementation of MS-SSIM

### ISSUE THREAD

**Issue by kunaltyagi** [kornia/kornia#1850]
## 🚀 Feature

Current implementation impacts training speed significantly (~40% slowdown compared to pure L1)

## Motivation

MS-SSIM gives some awesome results, and it'd be nice to have an option to switch to a 
* lower accuracy and faster implementation such as https://link.springer.com/article/10.1007/s11554-010-0170-9
* more optimized implementation

## Pitch

Faster iteration is always great. Difference between workflow for a task which starts in morning and ends during lunch compared to one that ends in evening is huge.

## Alternatives

Reduce the scales in MS-SSIM

## Additional context

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

**Comment by edgarriba** [comment]
@kunaltyagi do you have any suggestion for improvements ? I don't think your statement is fair to compare against L1 since the ssim/msssim is a more complex operator than element-wise subtraction and absolute value. But I agree, let's iterate to optimize if possible.

The way is implemented is via 2d convolutions and the reason for being slow might be because of large kernels (computed from sigmas). The default values gives the last a sigma of  8.0 which makes the max a kernel size of 32 times the number of scales (by default 5) times channels if i understand well the implementation. Definitely a large kernel.

https://github.com/kornia/kornia/blob/master/kornia/losses/ms_ssim.py

**Comment by kunaltyagi** [comment]
I have a couple of ideas, but I don't know how much they are worth since my experience in PyTorch doesn't include bench-marking.
1. Use integral values. They are usually fast than float (though does it matter in pytorch?)
2. Reduce kernel size -> Will have an immediate impact
3. Use jit or compile the kernel to allow optimizations based on the known values of kernel. Eg: multiplication with 5 can be implemented as a shifts + add but multiplication with unknown integer would be more expensive (generic muladd operation). Though some HW actually perform muladd faster than multiplication and shift (ask Intel)
4. Use other operation instead of expensive variance calculation 

Some of these are the same ideas as shown in the paper linked above. Apologies, but I can't find a PDF which I can distribute.

**Comment by edgarriba** [comment]
@kunaltyagi 
1) pytorch needs floating point values as requirements to differentiate.
2) We can explore and give a different default, however, as an api it should be customizable to enable more wide research. 
3) Not clear shifts implements the backward properly.
4) I believe variance is part of the formulation of ssim.

Tagging @t-vi  to contribute with ideas as for pytorch optimization here.

As for benchmark, if you want to try some of those ideas i suggest to read this tutorial
https://pytorch.org/tutorials/recipes/recipes/benchmark.html

Eventually, use this template: https://github.com/kornia/kornia/blob/master/test/benchmark.py

**Comment by dodobyte** [comment]
It is indeed pretty slow with large kernels which is expected. Current implementation uses regular convolution, but gaussian filter is separable, so using get_gaussian_kernel1d and filter2d_separable gives 3x speedup with the kernel size of 127.
